hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java

```diff
@@ -190,34 +192,71 @@ public DefaultMobStoreCompactor(Configuration conf, HStore store) {
     // Check if I/O optimized MOB compaction
     if (ioOptimizedMode) {
       if (request.isMajor() && request.getPriority() == HStore.PRIORITY_USER) {
-        Path mobDir =
-            MobUtils.getMobFamilyPath(conf, store.getTableName(), store.getColumnFamilyName());
-        List<Path> mobFiles = MobUtils.getReferencedMobFiles(request.getFiles(), mobDir);
-        //reset disableIO
-        disableIO.set(Boolean.FALSE);
-        if (mobFiles.size() > 0) {
-          calculateMobLengthMap(mobFiles);
+        try {
+          final SetMultimap<TableName, String> mobRefs = request.getFiles().stream()
+              .map(file -> {
+                byte[] value = file.getMetadataValue(HStoreFile.MOB_FILE_REFS);
+                ImmutableSetMultimap.Builder<TableName, String> builder;
+                if (value == null) {
+                  builder = ImmutableSetMultimap.builder();
+                } else {
+                  try {
+                    builder = MobUtils.deserializeMobFileRefs(value);
+                  } catch (RuntimeException exception) {
+                     throw new RuntimeException("failure getting mob references for hfile " + file,
+                         exception);
+                  }
+                }
+                return builder;
+              }).reduce((a, b) -> a.putAll(b.build())).orElseGet(ImmutableSetMultimap::builder)
+              .build();
+          //reset disableIO
+          disableIO.set(Boolean.FALSE);
```

Comment:
Does it matter if we don't do this in a finally block if we get the RTE from above?

Reply:
as far as I can tell, yes. we're setting it to `false` in prep for the call right after to `calculateMobLengthMap` where we check each of the file names to see if we need to switch the flag.

If instead we hit the RuntimeException from deserializing then we shouldn't make that call to `calculateMobLengthMap` and so we don't need to mess with the `disableIO` flag at all.